README
This is a sysinfo:/ KIO slave for KDE4, which shows basic
system information often requested by users.

Patched: plasma5 string + no more page reload

The code was originally from http://svn.opensuse.org/svn/kio_sysinfo

Patches and feedback greatly appreciated.

You can enable distro-specific paths with sth like:
cmake -DCMAKE_INSTALL_PREFIX=$KDEDIR -DSYSINFO_DISTRO:STRING=distro .

Currently supported distros are:
- fedora
- suse
- mandriva
- debian
- ubuntu

Dirk Mueller <dmueller@suse.de>
Lukas Tinkl <ltinkl@redhat.com>


debian/changelog
kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu14.04.6) utopic; urgency=low

  * Some fixes

 -- Leszek Lesner <leszek@zevenos.com>  Thu, 05 Feb 2015 18:44:05 +0100

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu14.04.5) utopic; urgency=low

  * Use the Plasma version when available
  * Disable reload every 20 seconds

 -- Leszek Lesner <leszek@zevenos.com>  Mon, 02 Feb 2015 18:39:40 +0100

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu14.04.4) trusty; urgency=low

  * Build against trusty

 -- Leszek Lesner <leszek@zevenos.com>  Fri, 25 Apr 2014 12:32:57 +0200

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu13.10.4) saucy; urgency=low

  * Updated style to Netrunne RC version one

 -- Leszek Lesner <leszek@zevenos.com>  Wed, 04 Dec 2013 17:04:15 +0100

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu13.10.3) saucy; urgency=low

  * Updated style again

 -- Leszek Lesner <leszek@zevenos.com>  Mon, 02 Dec 2013 11:39:19 +0100

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu13.10.2) saucy; urgency=low

  * Updated style.css

 -- Leszek Lesner <leszek@zevenos.com>  Tue, 05 Nov 2013 17:19:31 +0100

kio-sysinfo (0.1-0ubuntu1+f89087e~ubuntu13.10.1) saucy; urgency=low

  * Auto build.

 -- Rohan Garg <rohan16garg@gmail.com>  Wed, 16 Oct 2013 09:37:03 +0000

kio-sysinfo (0.1-0ubuntu1~ppa3) quantal; urgency=low

  * No change rebuild

 -- Rohan Garg <rohangarg@kubuntu.org>  Thu, 11 Oct 2012 06:10:41 +0530

kio-sysinfo (0.1-0ubuntu1~ppa2) quantal; urgency=low

  * Install file++

 -- Rohan Garg <rohangarg@kubuntu.org>  Thu, 04 Oct 2012 21:14:43 +0100

kio-sysinfo (0.1-0ubuntu1~ppa1) quantal; urgency=low

  * Initial release

 -- Rohan Garg <rohangarg@kubuntu.org>  Mon, 01 Oct 2012 21:37:34 +0100


src/sysinfo.h
//////////////////////////////////////////////////////////////////////////
// sysinfo.h                                                            //
//                                                                      //
// Copyright (C)  2005  Lukas Tinkl <lukas.tinkl@suse.cz>               //
//                                                                      //
// This program is free software; you can redistribute it and/or        //
// modify it under the terms of the GNU General Public License          //
// as published by the Free Software Foundation; either version 2       //
// of the License, or (at your option) any later version.               //
//                                                                      //
// This program is distributed in the hope that it will be useful,      //
// but WITHOUT ANY WARRANTY; without even the implied warranty of       //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        //
// GNU General Public License for more details.                         //
//                                                                      //
// You should have received a copy of the GNU General Public License    //
// along with this program; if not, write to the Free Software          //
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA        //
// 02110-1301, USA.                                                     //
//////////////////////////////////////////////////////////////////////////

#ifndef _sysinfo_H_
#define _sysinfo_H_

#include <qmap.h>
#include <qstringlist.h>

#include <kurl.h>
#include <kicontheme.h>
#include <kio/slavebase.h>

#include <solid/predicate.h>

#define GFX_VENDOR_ATI "ATI Technologies Inc."
#define GFX_VENDOR_NVIDIA "NVIDIA Corporation"

struct DiskInfo
{
    // taken from media:/
    QString id;
    QString name;
    QString label;
    QString deviceNode;
    QString mountPoint;
    QString fsType;
    bool mounted;
    bool removable;
    QString iconName;

    // own stuff
    quint64 total, avail; // space on device
};


/**
 * System information IO slave.
 *
 * Produces an HTML page with system information overview
 */
class kio_sysinfoProtocol : public KIO::SlaveBase
{
public:
    kio_sysinfoProtocol( const QByteArray &pool_socket, const QByteArray &app_socket );
    virtual ~kio_sysinfoProtocol();
    virtual void mimetype( const KUrl& url );
    virtual void get( const KUrl& url );

    /**
     * Info field
     */
    enum
    {
        MEM_TOTALRAM = 0,       // in sysinfo.mem_unit
        MEM_FREERAM,
        MEM_TOTALSWAP,
        MEM_FREESWAP,
        SYSTEM_UPTIME,          // in seconds
        CPU_MODEL,
        CPU_SPEED,              // in MHz
        CPU_CORES,
        CPU_TEMP,
        OS_SYSNAME,             // man 2 uname
        OS_RELEASE,
        OS_VERSION,
        OS_MACHINE,
        OS_USER,                // username
        OS_SYSTEM,              // OS version
        OS_HOSTNAME,
        GFX_VENDOR,              // Display stuff
        GFX_MODEL,
        GFX_2D_DRIVER,
        GFX_3D_DRIVER,
        BATT_IS_PLUGGED,        // see Solid::Battery
        BATT_CHARGE_PERC,
        BATT_CHARGE_STATE,
        BATT_IS_RECHARGEABLE,
        AC_IS_PLUGGED,          // see Solid::AcAdapter
        SYSINFO_LAST
    };

private:
    /**
     * Gather basic memory info
     */
    void memoryInfo();

    /**
     * Gather CPU info
     */
    void cpuInfo();

    /**
     * @return a formatted table with disk partitions
     */
    QString diskInfo();

    /**
     * Get info about kernel and OS version (uname)
     */
    void osInfo();

    /**
     * Gather basic OpenGL info
     */
    bool glInfo();

    /**
     * Gather battery / AC adapter status
     */
    bool batteryInfo();

    /**
     * Helper function to locate a KDE icon
     * @return img tag with full path to the icon
     */
    QString icon( const QString & name, int size = KIconLoader::SizeSmall ) const;

    /**
     * Fill the list of devices (m_devices) with data from the media KIO protocol
     * @return true on success
     */
    bool fillMediaDevices();

    /**
     * Map holding the individual info attributes
     */
    QMap<int, QString> m_info;

    QList<DiskInfo> m_devices;
    Solid::Predicate m_predicate;
};

#endif


src/ksysinfopart.h
/*  This file is part of the KDE project
    Copyright (C) 2002 Alexander Neundorf <neundorf@kde.org>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.
*/


#ifndef KSYSINFOPART_H
#define KSYSINFOPART_H

#include <kparts/factory.h>
#include <kparts/part.h>
#include <kparts/browserextension.h>
#include <khtml_part.h>
#include <kio/job.h>
#include <kio/jobclasses.h>
#include <kdirnotify.h>

//solid
#include <solid/device.h>

class KComponentData;
class KAboutData;

/**
 * Sysinfo Page Viewer
 * \todo: Why is it needed? Why is KHTML alone not possible?
 */
class KSysinfoPartFactory: public KParts::Factory
{
   Q_OBJECT
   public:
      KSysinfoPartFactory( QObject * parent = 0 );
      virtual ~KSysinfoPartFactory();

      virtual KParts::Part* createPartObject( QWidget * parentWidget,
                                QObject* parent, const char * classname,
                                const QStringList &args);

      static KComponentData * instance();

   private:
      static KComponentData * s_componentData;
      static KAboutData * s_about;

};

class KSysinfoPart : public KHTMLPart
{
   Q_OBJECT
   public:
      KSysinfoPart( QWidget * parent );

   protected slots:
      void onDeviceAdded(const QString &udi);
      void rescan();
      void slotResult( KJob *job );

   protected:
      KComponentData *m_instance;
      QTimer *rescanTimer;

#ifdef PORTED
      // Reimplemented from KDirNotify
      virtual void FilesAdded( const KUrl & dir );
      virtual void FilesRemoved( const KUrl::List & urls );
      virtual void FilesChanged( const KUrl::List & urls );
#endif

      void customEvent( QEvent *event );
      void mountDevice( Solid::Device *device );
      bool unmountDevice( Solid::Device *device );
      bool urlSelected( const QString &url,
                        int button,
                        int state,
                        const QString &_target,
                        const KParts::OpenUrlArguments &args = KParts::OpenUrlArguments(),
                        const KParts::BrowserArguments &browserArgs = KParts::BrowserArguments());
};

#endif



src/ksysinfopart.cpp
/*  This file is part of the KDE project
    Copyright (C) 2002 Alexander Neundorf <neundorf@kde.org>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.
*/

#include "ksysinfopart.h"

#include <qtimer.h>
#include <QMouseEvent>
#include <kcomponentdata.h>
#include <kglobal.h>
#include <kdebug.h>
#include <klocale.h>
#include <kstandarddirs.h>
#include <kaboutdata.h>
#include <kdeversion.h>
#include <kmenu.h>
#include <khtmlview.h>
#include <khtml_events.h>
#include <qcursor.h>
#include <kio/netaccess.h>
#include <kfileitem.h>
#include <KDesktopFile>

//solid
#include <solid/networking.h>
#include <solid/device.h>
#include <solid/deviceinterface.h>
#include <solid/devicenotifier.h>
#include <solid/predicate.h>
#include <solid/storagedrive.h>
#include <solid/storageaccess.h>
#include <solid/storagevolume.h>
#include <solid/opticaldisc.h>
#include <solid/opticaldrive.h>

extern "C"
{
    KDE_EXPORT void* init_libksysinfopart()
    {
        return new KSysinfoPartFactory;
    }
}

KComponentData* KSysinfoPartFactory::s_componentData = 0L;
KAboutData* KSysinfoPartFactory::s_about = 0L;

KSysinfoPartFactory::KSysinfoPartFactory( QObject* parent )
    : KParts::Factory( parent )
{}

KSysinfoPartFactory::~KSysinfoPartFactory()
{
    delete s_componentData;
    s_componentData = 0L;
    delete s_about;
}

KParts::Part* KSysinfoPartFactory::createPartObject( QWidget * parentWidget, QObject *,
                                                     const char* /*className*/,const QStringList & )
{
    KSysinfoPart* part = new KSysinfoPart(parentWidget);
    return part;
}

KComponentData* KSysinfoPartFactory::instance()
{
    if( !s_componentData )
    {
        s_about = new KAboutData( "ksysinfopart", 0, ki18n("KSysInfo"), KDE_VERSION_STRING,
                                  ki18n( "Embeddable System Information" ), KAboutData::License_GPL );
        s_componentData = new KComponentData( s_about );
    }
    return s_componentData;
}


KSysinfoPart::KSysinfoPart( QWidget * parent )
    : KHTMLPart( parent )
{
    KComponentData * instance = new KComponentData( "ksysinfopart" );
    setComponentData( *instance );
    rescanTimer=new QTimer(this);
    connect(rescanTimer, SIGNAL(timeout()), SLOT(rescan()));
    rescanTimer->setSingleShot(true);
    // Disable reloading as requested
    // rescanTimer->start(20000);
    setJScriptEnabled(false);
    setJavaEnabled(false);
    setPluginsEnabled(false);
    setMetaRefreshEnabled(false);

    connect(Solid::Networking::notifier(), SIGNAL(statusChanged(Solid::Networking::Status)),
            this, SLOT(rescan()));
    connect(Solid::DeviceNotifier::instance(), SIGNAL(deviceAdded(const QString &)),
            this, SLOT(onDeviceAdded(const QString &)));
    connect(Solid::DeviceNotifier::instance(), SIGNAL(deviceRemoved(const QString &)),
            this, SLOT(rescan()));

    QList<Solid::Device> deviceList = Solid::Device::listFromQuery("IS StorageAccess");
    Q_FOREACH (const Solid::Device &device, deviceList)
    {
        const Solid::StorageAccess *access = device.as<Solid::StorageAccess>();
        connect(access, SIGNAL(accessibilityChanged(bool, const QString &)),
                this, SLOT(rescan()));
    }

    installEventFilter( this );
}

void KSysinfoPart::slotResult( KJob *job)
{
    KIO::StatJob *sjob = dynamic_cast<KIO::StatJob*>(job);
    if (!job)
        return;

    KFileItemList list;
    list.append(KFileItem(sjob->statResult(), sjob->url()));
    emit browserExtension()->popupMenu(QCursor::pos(), list); //, );
}

void KSysinfoPart::customEvent( QEvent *event )
{
    if ( KParts::Event::test(event, "khtml/Events/MousePressEvent"))
    {
        khtml::MousePressEvent *ev = static_cast<khtml::MousePressEvent *>( event );
        KUrl url(ev->url().string());
        if (url.path().startsWith("/dev/") && (ev->qmouseEvent()->button() & Qt::RightButton))
        {
            KIO::UDSEntry entry;
            KIO::Job *job = KIO::stat(url, KIO::HideProgressInfo);
            connect( job, SIGNAL( result( KJob * ) ),
                     SLOT( slotResult( KJob * ) ) );
            return;
        }
    }

    KHTMLPart::customEvent(event);
}

void KSysinfoPart::mountDevice( Solid::Device *device )
{
    QStringList interestingDesktopFiles;
    foreach (const QString &path, KGlobal::dirs()->findAllResources("data", "solid/actions/")) {
        KDesktopFile cfg(path);
        const QString string_predicate = cfg.desktopGroup().readEntry("X-KDE-Solid-Predicate");
        QString fileUrl = KUrl(path).fileName();
        Solid::Predicate predicate = Solid::Predicate::fromString(string_predicate);
        if (predicate.matches(*device)) {
            interestingDesktopFiles << fileUrl;
        }
    }
    QDBusInterface soliduiserver("org.kde.kded", "/modules/soliduiserver", "org.kde.SolidUiServer");
    QDBusReply<void> reply = soliduiserver.call("showActionsDialog", device->udi(), interestingDesktopFiles);
}

bool KSysinfoPart::unmountDevice( Solid::Device *device )
{
    if (device->is<Solid::OpticalDisc>()) {
        Solid::OpticalDrive *drive = device->parent().as<Solid::OpticalDrive>();
        if (drive) {
            drive->eject();
        }
    }
    else if (device->is<Solid::StorageVolume>()) {
        Solid::StorageAccess *access = device->as<Solid::StorageAccess>();
        if (access && access->isAccessible()) {
            access->teardown();
            return true;
        }
    }

    return false;
}

bool KSysinfoPart::urlSelected( const QString &url,
                                int button,
                                int state,
                                const QString &_target,
                                const KParts::OpenUrlArguments& _args,
                                const KParts::BrowserArguments& _browserArgs)
{
    if (url.startsWith("#unmount=")) {
        Solid::Device device = Solid::Device(url.mid(9));
        if (!device.isValid()) {
            kDebug(1242) << "Device for udi" << device.udi() << "is invalid";
            return false;
        }
        return unmountDevice(&device);
    }

    QString path = KUrl(url).path();
    if (path.startsWith("/dev/")) {
        Solid::Predicate predicate(Solid::DeviceInterface::Block, "device", path);
        QList<Solid::Device> devList = Solid::Device::listFromQuery(predicate, QString());

        if (!devList.empty()) {
            Solid::StorageAccess *access = devList[0].as<Solid::StorageAccess>();
            if (access && access->isAccessible()) { //mounted
                openUrl(KUrl(access->filePath()));
                return true;
            }
            else { //not mounted
                Solid::Device device = Solid::Device(devList[0]);
                Solid::StorageDrive *drive = device.parent().as<Solid::StorageDrive>();
                if ((drive && drive->isHotpluggable()) || device.is<Solid::OpticalDisc>()) {
                    mountDevice(&device);
                    return true;
                }
                kDebug(1242) << "Device" << path << "not mounted and not hotpluggable";
                return false;
            }
        }
        else {
            kDebug(1242) << "Device" << path << "not found";
            return false;
        }
    }

    return KHTMLPart::urlSelected(url, button, state, _target, _args, _browserArgs);
}

void KSysinfoPart::onDeviceAdded(const QString &udi) {
    Solid::Device device = Solid::Device(udi);
    Solid::StorageAccess *access = device.as<Solid::StorageAccess>();
    if (access) {
        connect(access, SIGNAL(accessibilityChanged(bool, const QString &)),
                this, SLOT(rescan()));
    }
    rescan();
}

void KSysinfoPart::rescan()
{
    openUrl(KUrl("sysinfo:/"));
    rescanTimer->stop();
    rescanTimer->start(20000);
}

#ifdef PORTED
void KSysinfoPart::FilesAdded( const KUrl & dir )
{
    if (dir.protocol() == "media")
    {
        rescanTimer->stop();
        rescanTimer->start(10, true);
    }
}

void KSysinfoPart::FilesRemoved( const KUrl::List & urls )
{
    for ( KUrl::List::ConstIterator it = urls.begin() ; it != urls.end() ; ++it )
        FilesAdded( *it );
}

void KSysinfoPart::FilesChanged( const KUrl::List & urls )
{
    // not same signal, but same implementation
    FilesRemoved( urls );
}
#endif

#include "ksysinfopart.moc"
